### PR TITLE
Optimize MetricName

### PR DESCRIPTION
**Pull request by filipochnik**
Hey! During some internal benchmarking I discovered that MetricName class can become a bottle neck if a lot of metric names a generated dynamically on the hot path. 
Looking at the code, there were a few things slowing it down:
- `String.replaceAllLiterally` is very slow prior to Java 9
- `String.replaceAll` compiles regex on each method call
- `MetricName.append` was doing some unnecessary work

Some basic benchmarks, ran on Mid 2015 MacBook Pro with 2.5GHz Core i7
```
$ java -version
java version "1.8.0_172"
Java(TM) SE Runtime Environment (build 1.8.0_172-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.172-b11, mixed mode)
```

```
Benchmark                                                                       (className)  Mode  Cnt     Score     Error  Units
BenchMetricName.appendOld                                                 java.util.HashMap  avgt    5  1249.212 ±  35.993  ns/op
BenchMetricName.appendOld           com.sumologic.benchmark.bench.metrics_scala.TestClassA$  avgt    5  1765.970 ± 337.021  ns/op
BenchMetricName.appendOptimized                                           java.util.HashMap  avgt    5   218.687 ±  41.918  ns/op
BenchMetricName.appendOptimized     com.sumologic.benchmark.bench.metrics_scala.TestClassA$  avgt    5   221.283 ±   6.072  ns/op
BenchMetricName.fromClassOld                                              java.util.HashMap  avgt    5  2704.233 ± 146.685  ns/op
BenchMetricName.fromClassOld        com.sumologic.benchmark.bench.metrics_scala.TestClassA$  avgt    5  3520.235 ± 624.656  ns/op
BenchMetricName.fromClassOptimized                                        java.util.HashMap  avgt    5   223.203 ±  37.580  ns/op
BenchMetricName.fromClassOptimized  com.sumologic.benchmark.bench.metrics_scala.TestClassA$  avgt    5   609.280 ±  22.542  ns/op
BenchMetricName.fromNameOld                                               java.util.HashMap  avgt    5   437.847 ±   3.476  ns/op
BenchMetricName.fromNameOld         com.sumologic.benchmark.bench.metrics_scala.TestClassA$  avgt    5   630.851 ±   6.655  ns/op
BenchMetricName.fromNameOptimized                                         java.util.HashMap  avgt    5     3.750 ±   0.021  ns/op
BenchMetricName.fromNameOptimized   com.sumologic.benchmark.bench.metrics_scala.TestClassA$  avgt    5     3.755 ±   0.014  ns/op
```

### CODE REVIEW

**Comment by erikvanoosten** [comment]
Hi @filipochnik, that is pretty awesome work!

We have always recommended not use dynamically generate metrics, however, I know this is not always reality :)

Unfortunately, I can not possible allow commons-lang as a dependency. I have seen too many compatibility problems with it. Are similar changes possible without commons-lang?

**Comment by erikvanoosten** [comment]
Upon closer inspection I see that only `StringUtils.replace` is used from commons-lang. Could you please just copy/paste that method?

**Comment by filipochnik** [comment]
> We have always recommended not use dynamically generate metrics, however, I know this is not always reality :)

Good point, will look at that next :)

> Upon closer inspection I see that only StringUtils.replace is used from commons-lang. Could you please just copy/paste that method?

Sure, no problem.

File: metrics-scala/src/main/scala/nl/grons/metrics4/scala/StringUtils.scala [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  /**
+    * Normalize dots and return the resulting [[String]].
+    *
+    * Normalizing dots means collapsing all adjacent occurrences of '.' to a single
+    * character and stripping leading and trailing '.'.
+    *
+    * @param s the [[String]] to normalize
+    * @return a normalized [[String]]
+    */
+  def normalizeDots(s: String): String = {
```
Comment (erikvanoosten) [inline metrics-scala/src/main/scala/nl/grons/metrics4/scala/StringUtils.scala]: Please rename to collapseDots.

File: metrics-scala/src/main/scala/nl/grons/metrics4/scala/MetricName.scala [hunk 2]
```diff
@@ -64,6 +75,18 @@ class MetricName private (val name: String) {
...
-  def append(names: String*): MetricName =
-    new MetricName((name.split('.') ++ names.filter(_ != null)).filter(_.nonEmpty).mkString("."))
+  def append(names: String*): MetricName = {
+    if (names.isEmpty) {
+      return this
+    }
+
+    val sb = new StringBuilder(name)
+    names.view
+      .filterNot(StringUtils.isEmpty)
```
[the remark this answers is not in the document]
Comment (filipochnik) [reply metrics-scala/src/main/scala/nl/grons/metrics4/scala/MetricName.scala]: Yeah, it's a bit odd, but I'm calling from 3 different places. I didn't want to repeat myself. I can inline it if you prefer that.

[the remark this answers is not in the document]
Comment (erikvanoosten) [reply metrics-scala/src/main/scala/nl/grons/metrics4/scala/MetricName.scala]: True, but do we really need the assertions in the replace method?

[the remark this answers is not in the document]
Comment (filipochnik) [reply metrics-scala/src/main/scala/nl/grons/metrics4/scala/MetricName.scala]: Hmm in the current use-case not really. But for it to be a complete replace function yes.
I can remove them and move `isEmpty` in line to `append`. Your call :)

[the remark this answers is not in the document]
Comment (erikvanoosten) [reply metrics-scala/src/main/scala/nl/grons/metrics4/scala/MetricName.scala]: A common assumption in Scala programs is that we do not pass `null`. I am happy to do the same here. We do still need to check for `text.isEmpty` and `searchString.isEmpty`. (I did a little test, if you pass an empty search there you get an OOM :) .)

So the following pre-condition check is sufficient:
```scala
    if (text.isEmpty || searchString.isEmpty) {
      return text
    }
```

[the remark this answers is not in the document]
Comment (filipochnik) [reply metrics-scala/src/main/scala/nl/grons/metrics4/scala/MetricName.scala]: Sure, fixed :)

File: metrics-scala/src/main/scala/nl/grons/metrics4/scala/MetricName.scala [hunk 3]
```diff
@@ -82,7 +82,8 @@ class MetricName private (val name: String) {
 
     val sb = new StringBuilder(name)
     names.view
-      .filterNot(StringUtils.isEmpty)
+      .filter(_ != null)
+      .filter(_.nonEmpty)
```
Comment (erikvanoosten) [inline metrics-scala/src/main/scala/nl/grons/metrics4/scala/MetricName.scala]: Its probably faster if we combine these.

```scala
 .filter(n => n != null && n.nonEmpty)
```

**Comment by erikvanoosten** [comment]
One little thing left. Nice stuff @filipochnik ! I was blown away by `collapseDots`, very nice optimizations there.

File: metrics-scala/src/main/scala/nl/grons/metrics4/scala/MetricName.scala (hunk 3 above)
Comment (filipochnik) [reply metrics-scala/src/main/scala/nl/grons/metrics4/scala/MetricName.scala]: Done :)

**Comment by filipochnik** [comment]
My pleasure :) Will you be releasing a new version with this change?

**Comment by erikvanoosten** [comment]
Yes, definitely!

**Comment by erikvanoosten** [comment]
Version 4.0.2 contains these changes.

**Pull request merged**